File: tools/ghidra/CreateIsrFunctions.java
// Create and name the 8 interrupt handlers for LPC1765 firmware
// @category LPC1765
// @menupath Tools.LPC1765.Create ISR Functions
// @description Create and name the Cortex-M3 ISR functions from the vector table

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.SourceType;

public class CreateIsrFunctions extends GhidraScript {

    @Override
    public void run() throws Exception {

        String[][] isrs = {
            { "0x1e4",  "WDT_IRQHandler"    },
            { "0x29a",  "TIMER0_IRQHandler" },
            { "0xff48", "TIMER2_IRQHandler" },
            { "0xff6c", "TIMER1_IRQHandler" },
            { "0xaf08", "UART3_IRQHandler"  },
            { "0xf9e8", "EINT1_IRQHandler"  },
            { "0xfa0a", "EINT2_IRQHandler"  },
            { "0xfa2c", "EINT3_IRQHandler"  },
        };

        for (String[] isr : isrs) {
            Address addr = toAddr(isr[0]);
            Function func = getFunctionAt(addr);
            if (func == null) {
                func = createFunction(addr, isr[1]);
            } else {
                func.setName(isr[1], SourceType.USER_DEFINED);
            }
            if (func == null) {
                println("FAILED: " + isr[1] + " @ " + addr);
            } else {
                println("OK: " + isr[1] + " @ " + addr);
            }
        }

        println("Done: 8 ISR functions");
    }
}

File: tools/ghidra/create_isr_functions.py
# Create and name the 8 interrupt handlers for LPC1765 firmware
# @category LPC1765
# @menupath Tools.LPC1765.Create ISR Functions
# @description Create and name the Cortex-M3 ISR functions from the vector table

from ghidra.program.model.symbol import SourceType
from ghidra.program.model.listing import CodeUnit

program = getCurrentProgram()
listing = program.getListing()
fm = program.getFunctionManager()
space = program.getAddressFactory().getDefaultAddressSpace()

# (offset, name, comment)  -- offsets have the Thumb bit already cleared
ISRS = [
    (0x0001E4, "WDT_IRQHandler",    "Watchdog timeout: clear WDTOF + increment timeout counter"),
    (0x00029A, "TIMER0_IRQHandler", "System tick: set 0x7A0 flag + increment counter (T0MR2=1999)"),
    (0x00FF48, "TIMER2_IRQHandler", "Output/PWM update (sine wave generation)"),
    (0x00FF6C, "TIMER1_IRQHandler", "Counter + divide-by-40 (frequency/period calc)"),
    (0x00AF08, "UART3_IRQHandler",  "Serial RX handler (parse IIR RX interrupts)"),
    (0x00F9E8, "EINT1_IRQHandler",  "External interrupt 1: clear EXTINT1 + set flags"),
    (0x00FA0A, "EINT2_IRQHandler",  "External interrupt 2: clear EXTINT2 + set flags"),
    (0x00FA2C, "EINT3_IRQHandler",  "External interrupt 3: clear EXTINT3 + set flags"),
]


def make_function(offset, name, comment):
    addr = space.getAddress(offset)
    # Ensure Thumb disassembly exists first (Cortex language => Thumb-2)
    disassemble(addr)
    func = fm.getFunctionAt(addr)
    if func is None:
        func = fm.createFunction(name, addr, None, SourceType.USER_DEFINED)
        if func is None:
            print("FAILED: %s @ %s" % (name, addr))
            return
    else:
        func.setName(name, SourceType.USER_DEFINED)
    print("OK: %s @ %s" % (name, addr))
    if comment:
        listing.setComment(addr, CodeUnit.PLATE_COMMENT, comment)


for off, name, cmt in ISRS:
    make_function(off, name, cmt)

print("Done: %d ISR functions" % len(ISRS))
